lib/dialects/abstract/query-generator/composition-group.js

```diff
@@ -0,0 +1,52 @@
+'use strict';
+
+const Composition = require('./composition');
+
+/*
+ * Holds multiple compositions or objects cohercible to compositions
+ *
+ * Useful for grouping and manipulating anonymous compositions. It
+ * holds compositions or single composition items.
+ *
+ * Concatenable methods.
+ */
+
+class CompositionGroup {
+  constructor(...args) {
+    this.compositions = args.slice(0);
+  }
+  get length() {
+    return this.compositions.length;
+  }
+  add(...compositions) {
+    Array.prototype.push.apply(this.compositions, compositions);
+
+    return this;
+  }
+  // Inserts a new composition(*spacer*) between existing compositions
+  space(spacer) {
+    if (this.compositions.length === 0) return this;
+
+    this.compositions = this.compositions.slice(1).reduce((a, group) => {
+      a.push(spacer, group);
+      return a;
+    }, [this.compositions[0]]);
+
+    return this;
+  }
+  merge(group) {
+    this.compositions = this.compositions.concat(group.compositions);
+
+    return this;
+  }
+  slice(...args) {
+    const newObj =  new this.constructor();
+    newObj.compositions = this.compositions.slice(...args);
+    return newObj;
+  }
+  toComposition() {
+    return new Composition(...this.compositions);
+  }
+}
+
+module.exports = CompositionGroup;
```

Comment:
Would it be possible you export these via `module.exports.CompositionGroup = CompositionGroup`, makes it easier to write typings without adding any extra work for you :)

Reply:
I moved them to the same file `composition.js` and they are exported as `module.exports.Class = Class` for easier typings.

The classes `Slot`, `Placeholder`, `Composition` and `CompositionGroup` are very connected to each other. The only reason they were in separate files was to use `module.exports = Class`, but it seems to be inconvenient for typing.

I also did put the classes `QueryProto`, `SelectProto`, `InsertProto` and `UpdateProto` together in the file `query-proto.js` and are exported as `module.exports.Class = Class`.